### ISSUE TITLE
Envoy should bind in IPv4/v6

### ISSUE THREAD

**Issue by valentin2105** [heptio/contour#48]
Hi,

I'm trying Contour (and envoy too) in my Kubernetes cluster, but I'm not able to make it work due of Envoy that bind in IPv4 only in the pod. 

It would be awesome if all services can bind in IPv6 too, by default ?
```
/ # netstat -lptn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:8000          0.0.0.0:*               LISTEN      -
tcp        0      0 127.0.0.1:8001          0.0.0.0:*               LISTEN      -
tcp        0      0 127.0.0.1:9001          0.0.0.0:*               LISTEN      1/envoy
tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      1/envoy
```

By looking the code, it seem that it need a small change in `internal/contour/json.go` (l147) :

```
		Listeners: []envoy.Listener{{
			Name:    "ingress_http",
-			Address: "tcp://0.0.0.0:8080", // TODO(dfc) should come from pod.hostIP
+			Address: "tcp://[::]:8080", // TODO(dfc) should come from pod.hostIP

			Filters: []envoy.Filter{
```

What do you think about that ?

**Comment by davecheney** [comment]
Good point. This should really be configurable from the command line.

**Comment by davecheney** [comment]
Bumping to 0.3

**Comment by valentin2105** [comment]
Thanks Dave.

**Comment by davecheney** [comment]
I'm working on this now. I will have a partial fix for 0.3 and i'll look at other improvements in the future.

**Comment by valentin2105** [comment]
I will try directly when the new Docker image will be available. 
Thanks @davecheney !

**Comment by davecheney** [comment]
#127 has been committed, it's available in gcr.io/heptio-images/contour:master now.

I'm leaving this issue open because I suspect that #127 wasn't enough. Listening on `--envoy-https-address=::` bound _only_ to IPv6 which I couldn't test on AWS. Binding to `0.0.0.0` may dual stack bind to `::` but it's probably machine specific and I don't have the resources to test this at the moment.

Please let me know how you go, i've moved the milestone to 0.4 in case I need to expeclitly create an ipv4 and ipv6 listener.

**Comment by valentin2105** [comment]
I have try to add the args `--envoy-https-address=::` on both Contour & Envoy image (from the `contour-deployment-rbac.yaml` file. And at each try I got Envoy that bind on IPv4 only.

Maybe I doing it wrong ?

**Comment by davecheney** [comment]
This is flag should go on the `contour serve` container
```
        - args:
          - serve
          - '--envoy-https-address=::'
          - --incluster
```

**Comment by valentin2105** [comment]
Humm,

I don't find where I'm wrong : 

```
      - args:
        - serve
        - '--envoy-https-address=::'
        - --incluster
        image: gcr.io/heptio-images/contour:master
        imagePullPolicy: Always
        name: contour
```
```
14:48:36 › k -n heptio-contour exec -it contour-5964cb8f9-ljzww -c contour sh                                                                                                                 
/ # netstat -lptn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:9001          0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      -
tcp        0      0 127.0.0.1:8000          0.0.0.0:*               LISTEN      1/contour
tcp        0      0 127.0.0.1:8001          0.0.0.0:*               LISTEN      1/contour
```

**Comment by davecheney** [comment]
Can you pls check that k8s has picked up the change to the deployment
object.

**Comment by davecheney** [comment]
Here's an updated `02-contour.yml` file to replace the one in `deployment-grpc-v2`


[02-contour.yaml.zip](https://github.com/heptio/contour/files/1605531/02-contour.yaml.zip)

**Comment by davecheney** [comment]
Or you can try as a diff
```
diff --git a/deployment/deployment-grpc-v2/02-contour.yaml b/deployment/deployment-grpc-v2/02-contour.yaml
index aae2e5a..e575c44 100644
--- a/deployment/deployment-grpc-v2/02-contour.yaml
+++ b/deployment/deployment-grpc-v2/02-contour.yaml
@@ -32,7 +32,7 @@ spec:
         imagePullPolicy: Always
         name: contour
         command: ["contour"]
-        args: ["serve", "--incluster"]
+        args: ["serve", "--incluster", "--envoy-https-address=::"]
         volumeMounts:
         - name: contour-config
           mountPath: /config
```

**Comment by valentin2105** [comment]
I deleted all, 

Created from your .yaml file , 

And still don't work : 

```
Valentin at AirBook.local in [~/Downloads]
15:21:21 › k -n heptio-contour get pod -o wide                                                                                                                                                
NAME                       READY     STATUS    RESTARTS   AGE       IP                           NODE
contour-848d9795ff-4lt7w   2/2       Running   0          19m       1404:f100:e:88:0:e:88:5      worker04.k8s.domain.tld
contour-848d9795ff-5t9qv   2/2       Running   0          19m       1404:f100:e:88:0:e:88:d0ac   worker01.k8s.domain.tld

Valentin at AirBook.local in [~/Downloads]
15:20:24 › k -n heptio-contour exec -it contour-848d9795ff-4lt7w -c contour sh                                                                                                                
/ # netstat -ltpn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:8000          0.0.0.0:*               LISTEN      1/contour
tcp        0      0 127.0.0.1:8001          0.0.0.0:*               LISTEN      1/contour
tcp        0      0 127.0.0.1:9001          0.0.0.0:*               LISTEN      -

15:21:03 › curl http://\[1404:f100:e:88:0:e:88:5\]:8000                                                                                                                                       
curl: (7) Failed to connect to 1404:f100:e:88:0:e:88:5 port 8000: Connection refused
15:21:03 › curl http://\[1404:f100:e:88:0:e:88:5\]:8001                                                                                                                                       
curl: (7) Failed to connect to 1404:f100:e:88:0:e:88:5 port 8001: Connection refused
15:21:03 › curl http://\[1404:f100:e:88:0:e:88:5\]:9001                                                                                                                                       
curl: (7) Failed to connect to 1404:f100:e:88:0:e:88:5 port 9001: Connection refused

15:23:30 › ping6 1404:f100:e:88:0:e:88:5                                                                                                                                                   ( ⎈ renaissance )
PING6(56=40+8+8 bytes )1404:f100:e:88:0:e:88:5::97d:f053:30c5 --> 1404:f100:e:88:0:e:88:5
16 bytes from 1404:f100:e:88:0:e:88:5, icmp_seq=0 hlim=61 time=0.931 ms
16 bytes from 1404:f100:e:88:0:e:88:5, icmp_seq=1 hlim=61 time=0.776 ms
16 bytes from 1404:f100:e:88:0:e:88:5, icmp_seq=2 hlim=61 time=1.005 ms

**Comment by davecheney** [comment]
Can you please paste the output of

kubectl get -n heptio-contour po contour-848d9795ff-4lt7w -o yaml

**Comment by valentin2105** [comment]
Sure : 

```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/created-by: |
      {"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"heptio-contour","name":"contour-848d9795ff","uid":"3f5704f8-f1cd-11e7-ad52-86f02152f53c","apiVersion":"extensions","resourceVersion":"18752664"}}
  creationTimestamp: 2018-01-05T04:02:29Z
  generateName: contour-848d9795ff-
  labels:
    app: contour
    pod-template-hash: "4048535199"
  name: contour-848d9795ff-4lt7w
  namespace: heptio-contour
  ownerReferences:
  - apiVersion: extensions/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: contour-848d9795ff
    uid: 3f5704f8-f1cd-11e7-ad52-86f02152f53c
  resourceVersion: "18752841"
  selfLink: /api/v1/namespaces/heptio-contour/pods/contour-848d9795ff-4lt7w
  uid: 3f58320a-f1cd-11e7-ad52-86f02152f53c
spec:
  affinity:
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - podAffinityTerm:
          labelSelector:
            matchLabels:
              app: contour
          topologyKey: kubernetes.io/hostname
        weight: 100
  containers:
  - args:
    - -c
    - /config/contour.yaml
    - --service-cluster
    - cluster0
    - --service-node
    - node0
    - -l
    - info
    command:
    - envoy
    image: docker.io/envoyproxy/envoy-alpine:latest
    imagePullPolicy: Always
    name: envoy
    ports:
    - containerPort: 8080
      name: http
      protocol: TCP
    - containerPort: 8443
      name: https
      protocol: TCP
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /config
      name: contour-config
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: contour-token-wt42k
      readOnly: true
  - args:
    - serve
    - --incluster
    - '--envoy-https-address=::'
    command:
    - contour
    env:
    - name: KUBERNETES_SERVICE_HOST   (THIS IS TO TALK to IPv6-ONLY API) 
      value: k8s-api.domain.tld
    - name: KUBERNETES_SERVICE_PORT_HTTPS
      value: "6443"
    - name: KUBERNETES_SERVICE_PORT
      value: "6443"
    image: gcr.io/heptio-images/contour:master
    imagePullPolicy: Always
    name: contour
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /config
      name: contour-config
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: contour-token-wt42k
      readOnly: true
  dnsPolicy: ClusterFirst
  initContainers:
  - args:
    - bootstrap
    - /config/contour.yaml
    command:
    - contour
    image: gcr.io/heptio-images/contour:master
    imagePullPolicy: Always
    name: envoy-initconfig
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /config
      name: contour-config
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: contour-token-wt42k
      readOnly: true
  nodeName: worker04.k8s.domain.tld
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: contour
  serviceAccountName: contour
  terminationGracePeriodSeconds: 30
  volumes:
  - emptyDir: {}
    name: contour-config
  - name: contour-token-wt42k
    secret:
      defaultMode: 420
      secretName: contour-token-wt42k
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2018-01-05T04:02:34Z
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: 2018-01-05T04:03:32Z
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: 2018-01-05T04:02:29Z
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://2899eb529353ac5a922830a23252cd34a406334f5abf785d2373b541a607b021
    image: gcr.io/heptio-images/contour:master
    imageID: docker-pullable://gcr.io/heptio-images/contour@sha256:d434e812116ed44af3fa1a1e47c8212ffd0ace569cb79cf4c6d3e108e1b1a58d
    lastState: {}
    name: contour
    ready: true
    restartCount: 0
    state:
      running:
        startedAt: 2018-01-05T04:03:31Z
  - containerID: docker://9e5f33bbc0ded8bff4ce18b09fed651db95c6bd2c0345f776c0f84defb2c5964
    image: envoyproxy/envoy-alpine:latest
    imageID: docker-pullable://envoyproxy/envoy-alpine@sha256:bdb04e7291b887cf9121f1ed14b6fad0b8414b1321fff0153cef0d07c219abef
    lastState: {}
    name: envoy
    ready: true
    restartCount: 0
    state:
      running:
        startedAt: 2018-01-05T04:03:28Z
  hostIP: 10.244.20.204
  initContainerStatuses:
  - containerID: docker://953dcaf2f74b8613e58c5f22b2d175928933daeace9b04af1c76a632afb6239b
    image: gcr.io/heptio-images/contour:master
    imageID: docker-pullable://gcr.io/heptio-images/contour@sha256:d434e812116ed44af3fa1a1e47c8212ffd0ace569cb79cf4c6d3e108e1b1a58d
    lastState: {}
    name: envoy-initconfig
    ready: true
    restartCount: 0
    state:
      terminated:
        containerID: docker://953dcaf2f74b8613e58c5f22b2d175928933daeace9b04af1c76a632afb6239b
        exitCode: 0
        finishedAt: 2018-01-05T04:02:33Z
        reason: Completed
        startedAt: 2018-01-05T04:02:33Z
  phase: Running
  podIP: 1404:f100:e:88:0:e:88:5
  qosClass: BestEffort
  startTime: 2018-01-05T04:02:29Z
```

**Comment by davecheney** [comment]
Hmm, that looks right. This change affects the address envoy binds to, not contour. You probably want to run that exec command in the envoy container, not the contour container.

**Comment by valentin2105** [comment]
I'm don't seen How I put that to Envoy in that block : 

```
  containers:
  - args:
    - -c
    - /config/contour.yaml
    - --service-cluster
    - cluster0
    - --service-node
    - node0
    - -l
    - info
    command:
    - envoy
    image: docker.io/envoyproxy/envoy-alpine:latest

```

**Comment by davecheney** [comment]
It's a little confusing how the configuration generated by contour affects envoy.

Envoy starts with no listeners open. It connects to contour to receive all of its configuration. If there are vhosts to handle it will bind to http and/or https as required.

You can check the listeners by running something like this command in the envoy container.
```
/ # wget -O - http://127.0.0.1:9001/listeners | cat
Connecting to 127.0.0.1:9001 (127.0.0.1:9001)
-                    100% |*******************************************************************************************************************************************************************************************|    28   0:00:00 ETA
["[::]:8443","0.0.0.0:8080"]/ # 
```

**Comment by valentin2105** [comment]
I got nothing for now, But I will try to serve any vhost, thanks for tricks . 
```
Connecting to 127.0.0.1:9001 (127.0.0.1:9001)
-                    100% |***************************************************************************************************************************************************************|[]     2   0:00:00 ETA
```

**Comment by davecheney** [comment]
don't forget the `| cat` on the end

**Comment by valentin2105** [comment]
```
/ # wget -O - http://127.0.0.1:9001/listeners | cat
Connecting to 127.0.0.1:9001 (127.0.0.1:9001)
-                    100% |***************************************************************************************************************************************************************|[]     2   0:00:00 ETA
/ # 
```

**Comment by davecheney** [comment]
yes, if you have zero ingress objects in your cluster, then envoy will not open any listeners.

**Comment by valentin2105** [comment]
Of course !! Sorry for my noobness, It's first time with Envoy 😄 .

I will try theses Ingresses.

**Comment by davecheney** [comment]
it's ok, this stuff is still super new. I'll make sure that the lack of listeners is documented for 0.3

**Comment by davecheney** [comment]
Marking as closed as #132 is complete.
